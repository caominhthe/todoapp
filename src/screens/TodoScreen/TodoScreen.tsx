import React, { useCallback, useState } from 'react';
import {
  FlatList,
  KeyboardAvoidingView,
  Platform,
  SafeAreaView,
  StyleSheet,
  View,
  Text,
} from 'react-native';
import { useDispatch, useSelector } from 'react-redux';
import { RootStackScreenProps, ScreenNames } from '@routers';
import { TodoInputGroup, TodoItem } from '@components';
import { selectTodoEditMode, selectTodos } from '@selectors';
import { Todo } from '@types';
import {
  createTodo,
  deleteTodos,
  setTodoMode,
  updateTodo,
} from '@root/src/reducers';
import { Palette, spacing, typography } from '@theme';

type TodoScreenProps = RootStackScreenProps<ScreenNames.Todo>;

export const TodoScreen: React.FC<TodoScreenProps> = () => {
  const [selectedTodo, setSelectedTodo] = useState<Todo | null>();
  const [content, setContent] = useState('');
  const mode = useSelector(selectTodoEditMode);
  const todoList: Todo[] = useSelector(selectTodos);
  const dispatch = useDispatch();

  //Clear input content Items after create new todo item
  const onCreateTodo = useCallback(() => {
    if (content !== '') {
      dispatch(createTodo(content.trim()));
      setSelectedTodo(null);
      setContent('');
    }
  }, [content, dispatch]);

  //Clear input content and selected Items after Update new todo item
  const onUpdateTodo = useCallback(() => {
    if (selectedTodo && content !== '') {
      const updatedTodo: Todo = {
        ...selectedTodo,
        title: content.trim(),
      };
      setContent('');
      setSelectedTodo(null);
      dispatch(updateTodo(updatedTodo));
    }
  }, [selectedTodo, content, dispatch]);

  // Input turn into edit mode when click on an item, selected item will be highlight
  const onPressItem = useCallback(
    (todo: Todo) => {
      setSelectedTodo(todo);
      setContent(todo.title);
      dispatch(setTodoMode('edit'));
    },
    [dispatch, mode],
  );

  // Reset content input and mode after delete to avoid confuse when edit
  const onPressDelete = useCallback((todo: Todo) => {
    setContent('');
    dispatch(deleteTodos(todo.id));
  }, []);

  const renderItem = (todo: Todo) => {
    return (
      <TodoItem
        todo={todo}
        onPress={onPressItem}
        onDelete={onPressDelete}
        isSelected={todo.id === selectedTodo?.id}
      />
    );
  };

  return (
    <SafeAreaView style={styles.container}>
      <KeyboardAvoidingView
        behavior={Platform.OS === 'ios' ? 'padding' : 'height'}
        style={styles.keyboardAvoidView}
        keyboardVerticalOffset={100}
      >
        <FlatList<Todo>
          showsVerticalScrollIndicator={false}
          data={todoList}
          keyExtractor={(todo) => String(todo.id)}
          renderItem={({ item }) => renderItem(item)}
          contentContainerStyle={styles.flatListContent}
        />
        <View style={styles.inputContainer}>
          <TodoInputGroup
            value={content}
            onChangeText={(text) => setContent(text)}
            onPressButton={mode === 'add' ? onCreateTodo : onUpdateTodo}
            button={
              <Text style={styles.inputButton}>
                {mode === 'add' ? 'Add' : 'Edit'}
              </Text>
            }
          />
        </View>
      </KeyboardAvoidingView>
    </SafeAreaView>
  );
};

const styles = StyleSheet.create({
  container: {
    flex: 1,
    alignItems: 'center',
    backgroundColor: Palette.white,
  },
  scrollView: {
    flex: 1,
  },
  flatListContent: {
    flexGrow: 1,
    paddingBottom: spacing.large,
  },
  keyboardAvoidView: {
    flex: 1,
    justifyContent: 'space-between',
  },
  inputContainer: {
    borderTopWidth: 1,
    borderColor: Palette.grey,
    paddingTop: spacing.small,
  },
  inputButton: {
    fontWeight: 'bold',
    fontSize: typography.fontSize.medium,
    backgroundColor: '#e5e5e5',
    padding: spacing.medium,
    borderRadius: 8,
    overflow: 'hidden',
  },
});
