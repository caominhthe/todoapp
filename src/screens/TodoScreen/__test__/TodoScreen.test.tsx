import { render } from '@testing-library/react-native';
import * as ReactRedux from 'react-redux';
import { TodoScreen } from '../TodoScreen';
import configureStore from 'redux-mock-store';

const mockStore = configureStore([]);
const mockNavigate = jest.fn();
const baseProps = {
  route: {},
  navigation: {
    navigate: mockNavigate,
  },
} as any;

describe('Given Authentication', () => {
  let store: any;

  it('should render correctly with todo items', () => {
    store = mockStore({
      todo: {
        todoList: {
          id: {
            id: 'id',
            title: 'test',
            createdAt: Date.now(),
          },
        },
      },
    });
    const { toJSON } = render(
      <ReactRedux.Provider store={store as any}>
        <TodoScreen {...baseProps} />
      </ReactRedux.Provider>,
    );
    expect(toJSON()).toMatchSnapshot();
  });
});
