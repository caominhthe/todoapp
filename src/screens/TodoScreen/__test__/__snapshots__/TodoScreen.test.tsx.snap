// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Given Authentication should render correctly with todo items 1`] = `
<RCTSafeAreaView
  emulateUnlessSupported={true}
  style={
    Object {
      "alignItems": "center",
      "backgroundColor": "#FFFFFF",
      "flex": 1,
    }
  }
>
  <View
    onLayout={[Function]}
    style={
      Array [
        Object {
          "flex": 1,
          "justifyContent": "space-between",
        },
        Object {
          "paddingBottom": 0,
        },
      ]
    }
  >
    <RCTScrollView
      contentContainerStyle={
        Object {
          "flexGrow": 1,
          "paddingBottom": 24,
          "paddingTop": 12,
        }
      }
      data={
        Array [
          Object {
            "createdAt": 1652712787252,
            "id": "id",
            "title": "test",
          },
        ]
      }
      disableVirtualization={false}
      getItem={[Function]}
      getItemCount={[Function]}
      horizontal={false}
      initialNumToRender={10}
      keyExtractor={[Function]}
      maxToRenderPerBatch={10}
      onContentSizeChange={[Function]}
      onEndReachedThreshold={2}
      onLayout={[Function]}
      onMomentumScrollEnd={[Function]}
      onScroll={[Function]}
      onScrollBeginDrag={[Function]}
      onScrollEndDrag={[Function]}
      removeClippedSubviews={false}
      renderItem={[Function]}
      scrollEventThrottle={50}
      showsVerticalScrollIndicator={false}
      stickyHeaderIndices={Array []}
      updateCellsBatchingPeriod={50}
      viewabilityConfigCallbackPairs={Array []}
      windowSize={21}
    >
      <View>
        <View
          onLayout={[Function]}
          style={null}
        >
          <View
            style={
              Array [
                Object {
                  "alignItems": "center",
                  "borderColor": "grey",
                  "borderRadius": 8,
                  "borderWidth": 1,
                  "flexDirection": "row",
                  "marginHorizontal": 24,
                  "marginTop": 8,
                  "minHeight": 40,
                  "paddingHorizontal": 24,
                  "paddingTop": 8,
                  "paddingVertical": 8,
                },
                false,
              ]
            }
          >
            <View
              accessible={true}
              focusable={true}
              onClick={[Function]}
              onResponderGrant={[Function]}
              onResponderMove={[Function]}
              onResponderRelease={[Function]}
              onResponderTerminate={[Function]}
              onResponderTerminationRequest={[Function]}
              onStartShouldSetResponder={[Function]}
              style={
                Object {
                  "flex": 1,
                  "opacity": 1,
                }
              }
              testID="todo-item"
            >
              <Text
                style={
                  Object {
                    "flexWrap": "wrap",
                    "fontSize": 16,
                    "fontWeight": "bold",
                    "paddingHorizontal": 12,
                  }
                }
                testID="item-title"
              >
                test
              </Text>
            </View>
            <View
              accessible={true}
              focusable={true}
              onClick={[Function]}
              onResponderGrant={[Function]}
              onResponderMove={[Function]}
              onResponderRelease={[Function]}
              onResponderTerminate={[Function]}
              onResponderTerminationRequest={[Function]}
              onStartShouldSetResponder={[Function]}
              style={
                Object {
                  "opacity": 1,
                }
              }
              testID="delete-button"
            >
              <Text
                style={
                  Object {
                    "color": "blue",
                    "fontSize": 16,
                  }
                }
              >
                Delete
              </Text>
            </View>
          </View>
        </View>
      </View>
    </RCTScrollView>
    <View
      style={
        Object {
          "borderColor": "#DADADA",
          "borderTopWidth": 1,
          "paddingTop": 8,
        }
      }
    >
      <View
        style={
          Object {
            "alignItems": "center",
            "flexDirection": "row",
            "paddingBottom": 4,
            "paddingHorizontal": 12,
            "width": 750,
          }
        }
      >
        <TextInput
          allowFontScaling={true}
          multiline={true}
          onChangeText={[Function]}
          placeholder="Add new todo"
          rejectResponderTermination={true}
          style={
            Object {
              "flex": 1,
              "flexShrink": 1,
              "fontSize": 20,
              "fontWeight": "700",
            }
          }
          underlineColorAndroid="transparent"
          value=""
        />
        <View
          accessible={true}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "opacity": 1,
            }
          }
          testID="submit-button"
        >
          <Text
            style={
              Object {
                "backgroundColor": "#e5e5e5",
                "borderRadius": 8,
                "fontSize": 16,
                "fontWeight": "bold",
                "overflow": "hidden",
                "padding": 12,
              }
            }
          >
            Edit
          </Text>
        </View>
      </View>
    </View>
  </View>
</RCTSafeAreaView>
`;
